B-Sockets-ServerTesting/sockets-servertesting.md

```diff
@@ -0,0 +1,298 @@
+than doing it through the web browser. Here, we try to send an HTTP request to
+the domain name `example.com` on port `80` (the default for HTTP web servers)
+for the URI `/index.html`. Note that we specify the `-C` with `netcat` so that
+newlines are `\r\n` rather than `\n` -- a requirement of the HTTP protocol. This
+flag may vary depending on `netcat` version.
+
+```console
+$ nc -C example.com 80
+GET /index.html HTTP/1.0    # GET == method; /index.html == request URI; HTTP/1.1 == version
+                            # blank line to specify end of request
+HTTP/1.0 200 OK             # start of HTTP response. HTTP/1.0 == version; 200 OK == response status
+Accept-Ranges: bytes        # header
+Content-Type: text/html     # more headers...
+-removed for brevity-
+                            # blank newline to indicate end of headers and start of file contents
+<!doctype html>             # HTML contents
+<html>
+<head>
+    <title>Example Domain</title>
+-removed for brevity-
+```
+
+## Testing your multi-server
+
+### Siege
+Siege is a command-line tool that allows you to benchmark your webserver using
+load testing. Given a few parameters, Siege gives you information about the
+number of successful transactions to your website, percent availability, the
+latency/throughput of your server, and more.
+
+To install siege, run the following command:
+
+```
+sudo apt install siege
+```
+
+To use siege with your webserver in HW3, run your server and test with the
+following command:
+
+```
+siege http://<hostname>:<port>/<url>
+```
+
+This will run for an infinite amount of time. When you
+<kbd>Ctrl</kbd>-<kbd>C</kbd> out of the command, a list of statistics will be
+outputted on your terminal.
+
+A better way to test with siege is using its options. The `-c` and `-r` options
+are particularly useful, as they allow you to specify the number of concurrent
+"users" and repetitions per user, respectively. For example, the following
+command will create 25 concurrent users that will each attempt to hit the server
+50 times, resulting in 1250 hit attempts:
+
+`siege -c 25 -r 50 http://<hostname>:<port>/<URI>`
+
+There are many other options, specified in the siege man page. These include
+`-t`, which specifies how long each user should run (as opposed to how many
+times), and `-f`, which specifies a file path that contains a list of URLs to
+test. 
+
+### Additional guidance on testing/benchmarking
+When grading, we're going to test your implementation using a mix of manual
+connections (e.g. using tools like netcat) and stress testers like siege.
+
+You should use `netcat` to make sure that basic functionality works and that you
+can indeed have more than 1 connection to the server at any given time. `netcat`
+is nice because it allows you to establish a connection and then prompts you for
+the data to send. You should also use `netcat` to test that your cleanup logic
+is correct, as you can control exactly when connections start/terminate.
```

Comment:
add something like..

your server should be resilient to client failures -- use netcat to simulate client disconnecting at various points during the transaction and make sure that your server can gracefully handle these scenarios